packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_pane_controller.dart

```diff
@@ -17,51 +15,45 @@ import 'charts/vm_chart_controller.dart';
 
 class MemoryChartPaneController extends DisposableController
     with AutoDisposeControllerMixin {
-  MemoryChartPaneController(this.mode, {bool? isDeviceAndroid})
+  MemoryChartPaneController(this.mode, {ChartData? data})
       : assert(
-          mode == DevToolsMode.connected || isDeviceAndroid != null,
-          'If application is not connected, isDeviceAndroid must be provided.',
-        ),
-        data = ChartData(isDeviceAndroid: isDeviceAndroid) {
-    unawaited(_init());
-  }
-
-  factory MemoryChartPaneController.fromJson(Map<String, dynamic> map) {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return MemoryChartPaneController(
-      DevToolsMode.offlineData,
-      isDeviceAndroid: false,
-    );
-  }
-
-  Map<String, dynamic> toJson() {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return {};
-  }
-
-  DevToolsMode mode;
-
-  final ChartData data;
+          mode == ControllerCreationMode.connected ||
+              (mode == ControllerCreationMode.offlineData &&
+                  data != null &&
+                  data.isDeviceAndroid != null),
+          '$mode, $data, ${data?.isDeviceAndroid}',
+        ) {
+    if (mode == ControllerCreationMode.connected) {
+      this.data = ChartData(mode: ControllerCreationMode.connected);
+      _chartConnection = ChartVmConnection(
+        this.data.timeline,
+        isAndroidChartVisible: isAndroidChartVisible,
+      );
+    } else {
+      this.data = data!;
+      _chartConnection = null;
+      _paused.value = false;
+      recomputeChartData();
+      _paused.value = true;
```

Comment:
since `recomputeChartData` is synchronous, we do not need to set _paused.value to false and true around this call. The UI thread will never be released to update listeners, and the default value for `_paused.value` is already true.

Reply:
Added comment.